Start replacing bash tests with D

marler8997 opened the pull request:
@JinShil 

Here's one way we could do it.

Pros:
* Windows users can now run these tests....after many hours I was never able to get these bash tests running on windows
* The test suite will now work on any platform that DMD supports regardless of the tools supported on that platform.

Cons:
* Tests take longer, but we are prebuilding the dshell library so the extra time is minimized, It appears to add about `.8` seconds per test right now.  See comparison below.
* It's more code to maintain. The `dshell` module adds about 300 lines of code in this initial implementation.
* Risk of changing test logic during the port that could cause tests to no longer cover the original cases they were written for.  Of course, this risk applies to any change to any codebase.  However, I think this risk is pretty low based on the difference between the BASH and DShell tests.  Usually it's just wrapping the commands in a `run()` function, most of the logic stays the same.

### Current Test Performance Comparison
Bash Tests:
```bash
# run the tests once so the test tools get built
$ ./run.d runnable/test_shared.sh runnable/test9377.sh runnable/depsprot.sh 
$ time ./run.d runnable/test_shared.sh runnable/test9377.sh runnable/depsprot.sh -f
unit_test_runner is already up-to-date
d_do_test is already up-to-date
sanitize_json is already up-to-date
 ... runnable/depsprot.sh
 ... runnable/test9377.sh
 ... runnable/test_shared.sh

real	0m0.440s
user	0m0.585s
sys	0m0.157s
```

DShell Tests:
```bash
# run the tests once so the test tools get built
$ ./run.d dshell/*.d
$ time ./run.d dshell/*.d -f
unit_test_runner is already up-to-date
sanitize_json is already up-to-date
d_do_test is already up-to-date
 ... dshell/test_shared.d
 ... dshell/test9377.d
 ... dshell/depsprot.d

real	0m1.532s
user	0m3.162s
sys	0m0.486s
```

Because of parallelization, the real time to run 3 tests went from about `.5` seconds to `1.5` seconds.  However, the "user" time went from `.5` seconds `3` seconds, which means in total it added about 2.5 seconds of "CPU time" for 3 tests, which is an average of `.8` seconds per test.  So we can estimate that on average, the added time to compile each dshell test is about `.8` seconds.  There are probably about 30 tests which could be converted, which would mean adding `24/CPU_COUNT` seconds to the build.  So with a 4-core machine, about 6 "real" seconds for all 30 tests, an 8-core would be about 3 "real" seconds and so on.

wilzbach commented on test/tools/d_do_test.d:
```diff
@@ -962,11 +965,76 @@ int runBashTest(string input_dir, string test_name)
     return process.wait();
 }
+
+int runDShellTest(string input_dir, string test_name, EnvData* envData,
```
```suggestion
int runDShellTest(string input_dir, string test_name, ref EnvData envData,
```

wilzbach commented on test/tools/d_do_test.d:
```diff
@@ -962,11 +965,76 @@ int runBashTest(string input_dir, string test_name)
+          writeln();
+    writeln("==============================");
+}
+void mkdirFor(string filename)
```
DStyle: short description of the function + one line of whitespace between functions

wilzbach commented on test/tools/dshell.d:
```diff
@@ -0,0 +1,150 @@
+/**
+A small library to help write D tests to replace our BASH tests.
```
```suggestion
A small library to help write D shell-like test scripts.
```

wilzbach commented on test/tools/d_do_test.d:
```diff
@@ -962,11 +965,76 @@ int runBashTest(string input_dir, string test_name)
+    }
+    mkdirFor(output_file);
+    auto file = File(output_file, "w");
+    file.write(output.data);
```
Why not `toFile`: https://dlang.org/library/std/stdio/to_file.html (or `std.file.write`)?

marler8997 replied:
I ended up re-writing it a bit.  Now instead of `execute` I use `spawnProcess`.  This is so that if you kill the test in the middle of it, you can still access all the output that was already printed, it doesn't get lost when the process dies or gets killed.

RazvanN7 reviewed (approved):
Looks good, although I think would should take into account using DMD as a library for this; that would avoid creating a new process for every test, making our test suite to run faster.

wilzbach commented on test/compilable/crlf_dshell.d:
```diff
@@ -0,0 +1,52 @@
+    src.appendFile(format("static assert(dstr == \"%s\");\n", `\nfoo\nbar\nbaz\n`));
+
+    const obj = buildPath(RESULTS_TEST_DIR, TEST_NAME ~ "a" ~ env.OBJ);
+    run([env.DMD, "-m"~env.MODEL, "-c", "-D", "-Dd" ~ RESULTS_TEST_DIR, "-of" ~ obj, src]);
```
Why not use your `dmd` struct from below?

marler8997 replied:
wasn't sure which one was better....what do you think?

wilzbach commented on test/compilable/crlf_dshell.d:
```diff
@@ -0,0 +1,52 @@
+    run([env.DMD, "-m"~env.MODEL, "-c", "-D", "-Dd" ~ RESULTS_TEST_DIR, "-of" ~ obj, src]);
+
+    remove(obj, buildPath(RESULTS_TEST_DIR, TEST_NAME ~ ".html"), src);
+    return 0;
```
Is this necessary? If there's an error, we throw an exception/error, no?

thewilsonator commented on test/compilable/crlf_dshell.d:
```diff
@@ -0,0 +1,52 @@
+    const src = buildPath(RESULTS_TEST_DIR, TEST_NAME ~ ".d");
+    
+    src.createFile(
+          "#!/usr/bin/env dmd -run\r\n"
```
please make this a single `q{}` string.

marler8997 replied:
It's got to have `\r\n` line endings....I don't think `q{}` strings guarantee that

wilzbach replied:
That won't work because `\r` isn't a valid D token and I think the whole point of this test is to test for `\r` and friends.

thewilsonator replied:
ew.

wilzbach reviewed (commented):
Nice!

jacob-carlborg commented:
If you’re going to rewrite/replace tests I suggest you replace them with this style of tests: https://github.com/dlang/dmd/tree/master/test/unit

JinShil commented on test/compilable/imports/ddoc_any_postscript.d:
```diff
@@ -0,0 +1,10 @@
+module imports.ddoc_any_postscript;
```
I strongly recommend only converting one test per PR.  I could be wrong, but I suspect if this PR gets too complicated, it's going to become too controversial and kill the whole idea.

marler8997 replied:
yeah, I'll leave it at just the 2 for now.

JinShil commented:
This has a lot more infrastructure than what I envisioned.  I just wanted the test runner to run `rdmd /test/compilable/whatever.d` and ensure it had an exit code of 0.  I like it though.

JinShil commented:
> If you’re going to rewrite/replace tests I suggest you replace them with this style of tests

@jacob-carlborg Could you please elaborate?

JinShil commented on test/tools/d_do_test.d:
```diff
@@ -610,6 +610,9 @@ int tryMain(string[] args)
     string output_file    = result_path ~ input_file ~ ".out";
     string test_app_dmd_base = output_dir ~ envData.sep ~ test_name ~ "_";
 
+    if (test_base_name.endsWith("_dshell.d"))
```
I recommend putting these the .d files in a different directory instead.

marler8997 replied:
done

jacob-carlborg commented:
> @jacob-carlborg Could you please elaborate?
EDIT: Maybe the unit directory needs a readme.

@JinShil There's already a readme: https://github.com/dlang/dmd/blob/master/test/README.md#types-of-tests. Please let me know if you need any more information.

JinShil commented:
> Please let me know if you need any more information.

Ok, I read that, but how would one use that framework to rewrite the shell scripts in D.  It seems those tests are intended to verify the compiler's internal state.  Since you have familiarity with that framework, could you choose a shell script to convert and make PR to illustrate the idea?

jacob-carlborg commented:
> Ok, I read that, but how would one use that framework to rewrite the shell scripts in D. It seems those tests are intended to verify the compiler's internal state. Since you have familiarity with that framework, could you choose a shell script to convert and make PR to illustrate the idea?

It highly depends on what the shell script is doing. Not all tests make sense to convert to this new framework, like runnables, for example. I'll have to take a look at the existing ones to see what they're actually doing.

jacob-carlborg commented:
> Ok, I read that, but how would one use that framework to rewrite the shell scripts in D. It seems those tests are intended to verify the compiler's internal state. Since you have familiarity with that framework, could you choose a shell script to convert and make PR to illustrate the idea?

@JinShil In this particular test (assuming I understand the test correctly) it looks like it just lexes/parses a file with CRLF and mixed line endings. That sounds quite straight forward to me. Just put a string literal with some D code and CRLF line endings in a test and pass the string to `parseModule`, something like this:

https://github.com/dlang/dmd/blob/124b3a2767237a64b40f35afd2c93ee5fe7ebcd1/test/unit/self_test.d#L17-L31

jacob-carlborg commented:
@JinShil see https://github.com/dlang/dmd/pull/9849 for a rewrite of the crlf test to the new style.

JinShil commented:
> see #9849 for a rewrite of the crlf test to the new style.

Thank you @jacob-carlborg  I understand now.  I can see that it works for the crlf.sh shell test, but I don't think it won't work for all shell tests.  So, I still think something like what @marler8997 has created here will still be needed.

For those tests that could be ported to either style, I would leave it up to the one doing the work to choose which one they prefer, unless there was a compelling reason to choose one or the other.

marler8997 commented:
Thanks @jacob-carlborg I didn't know about this new unittest framework.  It seems a good fit for `crlf.sh`, I'll have to take a look the rest of the `.sh` tests to see what else will fit.

jacob-carlborg commented:
This test, `crlf.sh`, is a prefect example on why my framework is needed. What `crlf.sh` is doing is a poor man's test of the lexer. It's running the whole compiler, including code generation even though there's no need for that. Then it's using `static assert` to verify the value of tokens. What it should be doing is running only the lexer and then compare the set of generated tokens to the expected value.

My PR in https://github.com/dlang/dmd/pull/9849 is a direct port of the `crlf.sh` test. It would be better to do what I described above. But #9849 is a start.

jacob-carlborg commented:
> For those tests that could be ported to either style, I would leave it up to the one doing the work to choose which one they prefer, unless there was a compelling reason to choose one or the other.

If it's possible to use my framework I think that should be preferred. It will compile all the files into one executable and run it once instead of compiling hundreds (thousands?) of tiny executables and run them all. That will be slower. For this reason many of the existing tests are bundled into a single file make it unreadable. My framework scales better when new files are added. Sure, it might be slightly slower to compile if there's a lot of files. It's also possible to specify a single or multiple files to only test a few files. In addition to that it's also possible to filter on individual unit test blocks within a single file. The test by Jonathan is also making quite a lot of IO since the compiler is run as an external command it needs a file as input.

But there will be cases where my framework doesn't work, like everything in the `runnable` directory.

Basile-z commented on test/tools/dshell.d:
```diff
@@ -0,0 +1,171 @@
+    string DSEP;
+}
+__gshared TestEnv env;
+__gshared string TEST_DIR; // TEST_DIR should be one of compilable, fail_compilation or runnable
```
> TEST_DIR should be one of compilable, fail_compilation or runnable

So why not using a named `enum TEST_DIR : string {}` ?

marler8997 replied:
You could, but I'm not seeing the benefit...?

marler8997 replied:
I suppose that would let you do a final switch on it?

Though, I don't see anyone doing that with TEST_DIR...but I would see people doing that with other values...such as OS

Basile-z replied:
No it was just a question of principle. The initial comment was simply motivated by the fact that you seemed to say that the variable could only have 3 different values. So just forget this, this was not a request.

marler8997 commented:
Ok I've made some updates.

1. moved dshell tests into the `dshell` directory
2. the `dshell` library now supports BASH-LIKE variable expansion
3. I've removed the `crlf` test in favor of a couple from the `runnable` directory

marler8997 commented:
I had a go at replacing one of the tests with grep. I was able to add a few functions to emulate what grep was doing.  Check out `depsrot.d` and how it compares to the original `depsrot.sh`.

atilaneves commented:
I'm unsure of the benefit of converting existings tests, especially if they're slower that way. It's one thing to convert a test as part of fixing a bug, but refactoring for refactoring's sake is seldom a good time priority.

JinShil commented:
@atilaneves The primary reason for this is so we can do D development on Windows.  Currently people are having a hell of a time building the D repositories and running the test suite on Windows.  Part of the problem is the reliance on a Posix environment and Posix-like tools which Windows does not natively support.  If the build scriipts, tests and other infrastructure were all written in D, they would be portable to all platforms D supports and would be written in a common language that all D developers understand.

atilaneves commented:
I've never attempted to run the dmd tests on Windows so my experience there is limited. Is there any reason why using git bash, WSL or a docker container don't currently work?

JinShil commented:
> Tests take longer. Current, for tests that don't use grep (meaning std.regex), it adds about a second to compile the test. However, when you import std.regex, it adds about 4 to 5 seconds...yikes!

I'm sure that can be improved in time.  Although it may be more convenient, I doubt a full regex package is necessary.  Most text comparisons can be coded in plain old D.

> It's more code to maintain. The dshell module adds about 300 lines of code in this initial implementation

I don't thing all the infrastructure is necessary.  My original though was to just run `rdmd theTest.d` without any infrastructure and ensure exist code is 0.  Whatever utilities are needed for text comparisons, spawning processes, etc. can be coded into each test.... However....

...having a set of utilities already written and easily imported from a file is nice.  So although its 300 lines of code to maintain, it's leveraged over several files, so well-utilized.

At this point I'll take either one.

> Another solution that comes to mind is moving to another scripting tool that has good support on linux/mac/windows/bsd. Python is a candidate, but I'm sure there are others as well. Though, that still means we have a dependency on an external scripting language.

Please no.  D is sufficiently powerful to serve this purpose.  Users don't need yet another tool/language to install and learn to use when they have D.

JinShil commented:
> Is there any reason why using git bash, WSL or a docker container don't currently work?

I have't attempted WSL or git bash myself, because I don't see any reason to reach for those tools when we have already have D.  Even on Linux, when I have bash available, I still prefer to write my build scripts and other automation in D.  I find D a superior language to Bash, make, or any other language/tool, and I see no reason not to utilize it for automation.

wilzbach commented:
> Is there any reason why using git bash, WSL or a docker container don't currently work?

WSL and docker beat the purpose of running tests on Windows in the first place (`git bash` is somewhere in between and is currently required).

wilzbach commented on test/dshell/depsprot.d:
```diff
@@ -0,0 +1,19 @@
+        .grep("^$TEST_NAME.*${TEST_NAME}_private")
+        .grep("private")
+        .enforceMatches("Private import protection in dependency file should be 'private'");
+    rm(Vars.deps_file);
```
I know this is ported from the old tests, but there's actually no big reason to clean-up these temporary files. They are in the results directory anyhow and cleaning up individually takes time and sometimes even fails on Windows, so moving forward I would just not care about these temporary files, but eventually run dmd in a temporary directory which can then be purged. What do you think?

marler8997 replied:
Well I think I fixed the issue on windows: https://github.com/dlang/dmd/pull/9838

Not removing a test's output files would increase the disk-space requirement to run the entire test suite without cleaning in between.  Although, this may not be a concern depending on the size of the test artifacts.  If the test artifacts only take of 50% of the space of all the other files required to run the tests, then probably not a big concern.

The other concern is the added logic/places for error.  I think this can be fixed by instead putting all the output files in a sub-directory.  Then the logic to cleanup the tests could be moved into a common location that just cleans the whole subdirectory, which we could leave as an option if the user is concerned with disk space.  What about that?

marler8997 replied:
```bash
#Clean dmd repo:

[marler8997@nixos:~/ddev1/dmd]$ du -hs .
151M	.

#After making dmd:

[marler8997@nixos:~/ddev1/dmd]$ du -hs .
190M	.

#After running the full test suite:

[marler8997@nixos:~/ddev1/dmd]$ du -hs .
224M	.

```

marler8997 replied:
Ok, I've removed the `rm` calls in the tests themselves, and left in a comment in `d_do_test.d` that indicates how we would remove the files in between tests if we wanted:
```D
rmdirRecurse(testOutDir);
```

wilzbach commented on test/tools/d_do_test.d:
```diff
@@ -962,11 +966,94 @@ int runBashTest(string input_dir, string test_name)
     return process.wait();
 }
+
+int runDShellTest(string input_dir, string test_name, const ref EnvData envData,
```
Should have a short description comment

wilzbach commented on test/tools/d_do_test.d:
```diff
@@ -962,11 +966,94 @@ int runBashTest(string input_dir, string test_name)
+        glueSource.writeln(`module dshellglue;`);
+        glueSource.writefln(`enum TEST_DIR  = "%s";`, input_dir);
+        glueSource.writefln(`enum TEST_NAME = "%s";`, test_name);
+    }
```
Move in a separate function instead of creating a sub-scope?

marler8997 replied:
Meh, that means creatinga new function and passing at least 3 variables to it for 4 lines of code...

marler8997 replied:
I believe doing that would make the code-flow harder to understand. This may be a 100-line function but it has no loops/jumps, the code just flows from top to bottom. Just a few if statements with a max depth of 2 and a few blocks to separate some variables into sub-scopes. Having long functions is fine, it is more about what the function is doing I think.

By splitting the code into multiple functions now you have to start checking/verifying/understanding what the state is in each method. You can't assume that a function that should have run before unless you go through all the code and verify it is always called before you. It provides less guarantees as now if you require something to run before you, you now need to verify that thing is always called before you. All that being said, if the function doesn't require this then it's not really an issue, but having the code be in it's own function stills adds more strain to the reader since just having it in a function means it can be called at any time, where with a long function, the reader knows implicitly when that code is going to be executed in relation to everything else.

wilzbach replied:
FWIW I was thinking about sth. like:

```d
struct DShellTester {
   void run() {
     initialize();
     createDShellModules();
     compileTests();
     runTests();
     cleanupArtifacts();
   }
}
```

where the reader immediately gets an high-level overview of what's happening and can then dive into the function he is actually interested in. The fact that you already use scopes for grouping shows that visual grouping is necessary ;-)

marler8997 replied:
Ok here it is.  For me this is MUCH harder to follow.  It seems unnecessarily complex and convoluted.  The single function was much simpler and you could immediately tell what was going to be executed and in what order without having to understand multiple functions.  But maybe this one is easier for you to follow?

wilzbach commented on test/tools/d_do_test.d:
```diff
@@ -962,11 +966,94 @@ int runBashTest(string input_dir, string test_name)
+        writeln("==============================");
+        remove(output_file_temp);
+    }
+    return exitCode;
```
This is quite a big function. I recommend splitting it up into multiple methods or a struct with multiple members for better readability.

wilzbach commented on test/tools/dshell.d:
```diff
@@ -0,0 +1,290 @@
+public import std.stdio;
+public import std.process;
+
+struct Vars
```
description comment

wilzbach commented on test/tools/dshell.d:
```diff
@@ -0,0 +1,290 @@
+    static string opDispatch(string name)() { return get(name); }
+}
+
+alias requiredEnvVars = AliasSeq!(
```
description

wilzbach commented on test/tools/dshell.d:
```diff
@@ -0,0 +1,290 @@
+        assert(expanded !is null, "codebug");
+        map[name] = expanded;
+    }
+    static string get(string name)
```
description

wilzbach commented on test/tools/dshell.d:
```diff
@@ -0,0 +1,290 @@
+    }
+}
+
+string fatal(T...)(T args)
```
description

wilzbach commented on test/tools/dshell.d:
```diff
@@ -0,0 +1,290 @@
+    writeln("Error: ", args);
+    return null;
+}
+string requireEnv(string name)
```
description + whitespace

wilzbach commented on test/tools/dshell.d:
```diff
@@ -0,0 +1,290 @@
+    return result;
+}
+
+void rm(T...)(T args)
```
description + whitespace

wilzbach commented on test/tools/dshell.d:
```diff
@@ -0,0 +1,290 @@
+                } catch (Exception e) {
+                    if (attempt >= 4)
+                        throw e;
+                    Thread.sleep(dur!"seconds"(1));
```
how about starting with 10ms and doing an exponential back-off?

marler8997 replied:
done...except, I just did a "multiply by 2 each time" backup

wilzbach reviewed (changes requested):
I'm in favor of this as maintaining (and creating new shell tests) is very painful. Being able to use the power of D to express is a huge win (and we get full cross-platform compatibility for free).
However, as other people probably would have to improve this script, I would like to see a lot more documentation.

atilaneves commented:
@JinShil Everything you're saying makes sense in a world in which we have no scripts and are choosing D over bash. I'm saying that I don't think porting *existing* tests to D is very helpful. It's the same with refactoring: my belief is that it should be done in the service of fixing a bug or adding functionality, not because the code would be nicer.

wilzbach commented:
> I don't think porting existing tests to D is very helpful

1) We need to start somewhere (hence a few existing tests) otherwise new tests can't be written with this.
2) You currently can't run the testsuite out of the box on Windows (so this is a huge win)
3) We have been converting a lot of the arcane infrastructure recently with huge success (people can finally build dmd on Windows without needing to hack the makefiles). This is the logical next step.

> It's the same with refactoring: my belief is that it should be done in the service of fixing a bug or adding functionality, not because the code would be nicer.

Well, that's how you aggregate "code debt" for which the DMD codebase is a prime example. Hence, we constantly merge PRs which refactor the codebase. Basically 90% of Walter's work on DMD is refactoring:

![image](https://user-images.githubusercontent.com/4370550/58423110-93851b80-8094-11e9-9f78-f2593c0f9ee9.png)

![image](https://user-images.githubusercontent.com/4370550/58423201-cc24f500-8094-11e9-8079-91657ee211b1.png)


(from https://github.com/dlang/dmd/pulls?q=is%3Apr+label%3ARefactoring+is%3Aclosed and https://github.com/dlang/dmd/pulls?q=is%3Apr+is%3Aclosed+author%3AWalterBright).

I don't see why we shouldn't modernize the test infrastructure too.

thewilsonator commented:
> We have been converting a lot of the arcane infrastructure recently with huge success

Indeed, dealing with shell tests are a right PITA and I've been unable to wrangle my way through them in the past (I had to defer to Rainer I think), what with dealing with extra files, checking the output matches some known good, successfully removing generated files (its even more infuriating than dealing with bureaucratic processes).

We also need to dogfood some examples as a point of reference for new tests whose intent would be more easily maintained as a D script rather than a shell script.

atilaneves commented:
I'm not opposed to writing new tests in D. I'm aware that would require infrastructure, so I'm not opposed to that either. I am however opposed to migrating existing tests, with one of the reasons being that I consider it lateral movement. But there's a better one, and that's the fact that due to (fortunately!) the absence of "test tests", there's no way of really knowing if the D code corresponds 100% to the bash code it replaced. The likelihood of them breaking mid-transition is high. New tests in D wouldn't suffer from this.

I understand the need for Windows contributors to work on dmd, and obviously this means being able to run the tests. I think the way forward there is to write a script (batch, powershell or D) that enables an out-of-the-box experience running the tests that already exist. I say just use git bash for now. After all, that's how git got ported to Windows.

marler8997 commented:
> I'm unsure of the benefit of converting existings tests, especially if they're slower that way. It's one thing to convert a test as part of fixing a bug, but refactoring for refactoring's sake is seldom a good time priority.

I list the PROS/CONS in the description above. The "PROS" section would be the reasons for converting the tests.

> I've never attempted to run the dmd tests on Windows so my experience there is limited. Is there any reason why using git bash, WSL or a docker container don't currently work?

I've attempted to run the DMD test suite on windows many times over many years...and I have never once been successful.  I've posted many questions asking for help in the forums, create issues for someone to post a tutorial on how to do it, and still have never been able to run a single bash test on windows.

I've spent hours reading GNU documentation and trying to build the posix tools on windows, and was mostly met with a myriad of bugs and problems that after a long number of hours I just have to give up.

D's development support on windows is severely lacking.  Building the repositories on windows vs linux is a night and day difference.  I use to use windows exclusively to develop with DMD, and after I started using linux I couldn't believe how easy it was :)

WSL is a pretty neat and interesting tool.  Unfortunately though, it only runs on Windows 10.  And it's currently going through a big refactor where instead of emulating the kernel syscalls, they're going to run a real kernel in a lightweight VM.  So I'm not sure what's going to happen with that.  However if you're interested, I put up a nice tutorial on how to install your own distro's to WSL

https://github.com/marler8997/wsl-register

I can try to explain the frustration and problems with Windows, but I think the best way to understand these issues would be to try it for yourself.   When you have time, take a stab at building dmd/druntime/phobos on windows, and then try to run the DMD test suite.  Then let us know how it went.

wilzbach commented:
>  I think the way forward there is to write a script (batch, powershell or D) that enables an out-of-the-box experience running the tests that already exist. 

Atila, we already have a test suite runner written completely in D (https://github.com/dlang/dmd/blob/master/test/run.d). However, at the moment the bash tests fail on Windows (for obvious reasons).

Sure, we could download git bash in the background (that's what digger does), but people have suffered from this long enough and in my humbled opinion it's a step backwards. As other people have mentioned, writing bash isn't very intuitive and there have been number of cases where people added bash tests which were wrong and never actually tested what they intended to test. Hell, even Walter still struggles with our bash test, which alone should be a reason to get rid of them.

The only real con I can see here is that this might add increase the compilation time, however, if this ever occurs, there's are multiple simple fixes (e.g. all tests could be compiled into one binary).

wilzbach reviewed (approved):
I wanted to do this for quite a while and I'm happy that @marler8997 is helping out here.
FWIW Walter and Andrei approved moving our test infrastructure from arcane Makefile/Bash scripts to pure D a long time ago, so the only discussion required should be about the actual cons like increased compilation time.

marler8997 commented:
Another thing I could try is building the `dshell.d` module only once, rather than rebuilding that module for each test.  That might help build times.

atilaneves commented:
> > I think the way forward there is to write a script (batch, powershell or D) that enables an out-of-the-box experience running the tests that already exist.
> 
> Atila, we already have a test suite runner written completely in D (https://github.com/dlang/dmd/blob/master/test/run.d). However, at the moment the bash tests fail on Windows (for obvious reasons).

I don't understand how that's related to my comment. I'm saying it would be nice if a foolproof tool existed that just worked for the current bash tests. That is, if a new Windows developer came along the very next day, they'd run that command and it'd successfully execute all dmd bash tests.
 
> Sure, we could download git bash in the background (that's what digger does),

That seems to indicate the approach would work.

> but people have suffered from this long enough and in my humbled opinion it's a step backwards. As other people have mentioned, writing bash isn't very intuitive and there have been number of cases where people added bash tests which were wrong and never actually tested what they intended to test. Hell, even Walter still struggles with our bash test, which alone should be a reason to get rid of them.

I'm not advocating for bash. I'm advocating for not porting the tests that have already been written. I don't like bash either.

> The only real con I can see here is that this might add increase the compilation time, however, if this ever occurs, there's are multiple simple fixes (e.g. all tests could be compiled into one binary).

The biggest con is breaking tests by porting them.

atilaneves commented:
> I've spent hours reading GNU documentation and trying to build the posix tools on windows,

I've also tried that, to no avail. Fortunately we don't need to do that anymore.

> and was mostly met with a myriad of bugs and problems that after a long number of hours I just had to give up. And keep in mind, we're not just talking about bash, but a whole set of programs that need to run on windows:
> 
> ```
> mkdir ln mv cp rm which uname bash tr awk git wc cat dirname wc head diff grep date sed
> ```

All of those exist in git bash for Windows (I literally just typed all them one by one right now).

marler8997 commented:
> All of those exist in git bash for Windows (I literally just typed all them one by one right now).

Interesting, I will have a go at trying that tonight I think.

> The biggest con is breaking tests by porting them.

I'm not sure why this is a concern.  Tests can't be merged if they're broken.

atilaneves commented:
> > The biggest con is breaking tests by porting them.
> 
> I'm not sure why this is a concern. Tests can't be merged if they're broken.

What I mean is that it's possible (and therefore *will* happen) that in the process of porting a bash script to D one inadvertently ends up with a D test that passes but isn't the same as the original test. That is, one ends up with a test that doesn't catch the same regressions as the original. A fake green, if you will.

marler8997 commented:
@atilaneves Agreed, a valid concern.

marler8997 commented:
Ok I've updated the implementation to prebuilt the `dshell` library, so now the tests shouldn't take much longer than the bash tests.

jacob-carlborg commented:
> I'm unsure of the benefit of converting existings tests, especially if they're slower that way. It's one thing to convert a test as part of fixing a bug, but refactoring for refactoring's sake is seldom a good time priority.

If the bash scripts are converted to this style of tests [1] then it should be faster. Because all those tests are compiled all once together and run as a single executable. Compare that to the current tests which runs the compiler once per file, that will be hundreds (or thousands?) of invocations.

Although this new style of tests [1] will not work for anything in the `runnable` directory.

[1] https://github.com/dlang/dmd/pull/9849

JinShil commented:
>  I'm saying that I don't think porting existing tests to D is very helpful. It's the same with refactoring: my belief is that it should be done in the service of fixing a bug or adding functionality, not because the code would be nicer.

Making the code nicer is not the motivation here.  The "bug" being fixed here, is that the test suite cannot be run on all platforms with the same experience.  The motivation is to create the exact same development experience on all platforms supported by D, without requiring any special environment setup, and to do so using a tool/language that all D developers would have in common.

> The biggest con is breaking tests by porting them.

Agreed.  It's a valid concern, but it's one that exists for any and all D development.  Fixing bugs in the compiler sometimes introduces regressions.  It's not any worse with this effort,

> I say just use git bash for now.

I just tried git bash and it doesn't work.  The test suite started throwing errors at me so fast it took me 30 seconds of pressing Ctrl-C to get it to stop.  The errors seem to indicate that it can't find druntime.  I don't know how to solve that problem.  Maybe I need to install the DMD from withing git bash to make it work. Maybe I need to customize ldconfig to make it work.  I don't know.  I'd rather spend my time converting the existing shell scripts to D than investigate this problem.  Converting the shell scripts to D will permanently remove any incompatibility problems for all platforms D supports now and in the future.  It's a permanent, long-term solution.

The other problem with git bash is it's not the same Windows environment that users will be typically be running D programs in.  That means that if we run the shell tests in git bash we're not actually doing a very good job of testing D on Windows, because it's not the native Windows environment.  The same for WSL.

> I am however opposed to migrating existing tests

I [suggested](https://github.com/dlang/dmd/pull/9848#discussion_r286831769) to @marler8997 to keep the scope of this PR small.  This PR originated out of a recent PR that added a new shell test.  I suggested in that PR to write the new shell test in D.  That grew into this PR.

I was afraid expanding the scope of this PR would undermine the effort to make our test suite portable, and it appears my instincts were right.  I'd be OK with, as a policy, preferring new shell tests be written in D, but I strongly recommend that, if someone is willing to volunteer their time and effort to fix the portability problems in D's test suite by converting our shell scripts to D, we give them our support.

atilaneves commented:
I think there's a misunderstanding about git bash on Windows - it's Windows with a different shell, that's all it is. There's no "installing dmd within git bash". It picks up everything that's in `%PATH%`. It's not a separate environment.

I'm willing to do the work of getting the existing bash tests working on Windows.

marler8997 commented:
Hey @atilaneves 

I was actually able to run a couple bash tests on windows today with git bash on windows.  It does some strange things, such as emulating a virtual filesystem tree within the tools themselves.  My tools are installed to `C:\Program Files\Git\bin\*` but they appear as `/bin/*`.  So in a way, it is a separate environment in that you can't access those paths from other windows tools.  I'd like to understand more what these tools are doing...I don't know of a clear definition of how it handles things.

I'm quite happy that I was able to run some of the tests, though there are still problems. I updated the wiki a bit to help: https://wiki.dlang.org/DMD_development#Running_the_testsuite_on_Windows

(See "DMD Test Suite - Windows Requirements")

I think it's a great idea for you to try to run the DMD test suite on windows.  Please let us know how it goes and if you need any help along the way.  Sometimes I'm on slack as well.

JinShil commented:
> I'm willing to do the work of getting the existing bash tests working on Windows.

I'm sure you'll succeed, but that's beside the point.  Writing all of our automation in D is a permanent, long-term, portable solution using a language/tool that all D programmers understand, enjoy using, and is more powerful than any shell or build tool.  It requires no additional dependencies and will extend to any platform D supports.  It's simply a superior solution.

atilaneves commented:
> 

> Writing all of our automation in D is a permanent, long-term, portable solution using a language/tool that all D programmers understand, enjoy using, and is more powerful than any shell or build tool. It requires no additional dependencies and will extend to any platform D supports. It's simply a superior solution.

I agree. So let's write new integration tests in D. I still don't want to port the existing tests because

1) The existing ones work
2) Some of the ports will have bugs that won't be caught by CI
3) I think there are better uses of community time

Then there's whether or not one should even be invoking the compiler over and over again to test it instead of writing unit tests instead.

JinShil commented:
> So let's write new integration tests in D.   I still don't want to port the existing tests

If that's the only option available (and it appears that it is), the only rational thing for me to do is concede and accept.

It's unfortunate, though, that we have a talented contributor able and willing to improve our automation's implementation and portability, with much work already laid before us, yet we're simply not willing to accept it.

JinShil commented:
> Then there's whether or not one should even be invoking the compiler over and over again to test it instead of writing unit tests instead.

marler8997 commented:
> I agree. So let's write new integration tests in D. I still don't want to port the existing tests because

So would you like me to remove the 3 tests I ported from bash and restore the original bash tests?  And if this is what you want, do you want this PR to just add the framework for now but not use it?  Or would you want it to be included alongside another set of changes in another PR where we add new test?

For example, I could close this PR and move the changes to this PR : https://github.com/dlang/dmd/pull/8121/files

Or this PR: https://github.com/dlang/dmd/pull/9831

Then I would rewrite the new bash tests in those PRs to D before merging them. Or would you rather keep this a PR that adds the framework but doesn't use it yet?

Also, did you get a chance to try running the test suite on windows?

marler8997 commented:
@atilaneves another thing to consider is it looks like we've finished removing all the C++ code in DMD.  Very similar to what this PR suggests except we're going from BASH to D.  In fact, converting the backend to D has very similar pros and cons to converting the BASH scripts to D (except the performance reason).  But your main reason for not wanting to convert the BASH tests to D is risk of regression.  So in your eyes this risk is a big factor when considering translating languages. So in your opnion, do you think we should have stuck with C++ in the backend instead of translating it?

thewilsonator commented:
The C++ -> D had a test suite backing it, and was between two similar languages.

marler8997 commented:
Well, here's one of the 3 tests that were converted that @atilaneves claims is too high of a risk to keep.  You be the judge as to how different they are:
```bash
libname=${OUTPUT_BASE}${LIBEXT}

$DMD -m${MODEL} -I${EXTRA_FILES} -of${libname} -c ${EXTRA_FILES}${SEP}mul9377a.d ${EXTRA_FILES}${SEP}mul9377b.d -lib || exit 1
$DMD -m${MODEL} -I${EXTRA_FILES} -of${OUTPUT_BASE}${EXE} ${EXTRA_FILES}${SEP}multi9377.d ${libname} || exit 1
rm_retry ${OUTPUT_BASE}{${LIBEXT},${OBJ},${EXE}}
```

```D
import dshell;
void main()
{
    Vars.set("libname", "$OUTPUT_BASE/a${LIBEXT}");

    run("$DMD -m$MODEL -I$EXTRA_FILES -of$libname -c $EXTRA_FILES/mul9377a.d $EXTRA_FILES/mul9377b.d -lib");
    run("$DMD -m$MODEL -I$EXTRA_FILES -of$OUTPUT_BASE/a$EXE $EXTRA_FILES/multi9377.d $libname");
}
```

thewilsonator commented:
> "$OUTPUT_BASE/a${LIBEXT}"

Why the extra 'a'? Why is LIBEXT in braces but everything else not?

marler8997 commented:
The bash version was writing files to a shared directory, whereas with the D tests I had them write all their files to their own directory.  You may notice that the D version doesn't have an `rm` command to clean up the test files, this is because those files can now be removed by the tester because now it just needs to cleanup the tests's whole directory.

So a name was needed for the executable, `a` is short and seems to be a common name among other tests.  And `${LIBEXT}` doesn't need the braces.

thewilsonator commented:
> You may notice that the D version doesn't have an rm command to clean up the test files

Hallelujah! I truly despised that requirement of the shell tests.

marler8997 commented:
Took out the braces from `${LIBEXT}` for the sake of consistency.

WalterBright commented:
How can WIP be automerged?

WalterBright reviewed (dismissed):
Can't have both WIP and automerge labels.

thewilsonator commented:
@marler8997 is this still WIP?

marler8997 commented:
No, not WIP. I can't remove labels.

thewilsonator commented:
Done.

wilzbach commented on test/tools/dshell_prebuilt/dshell_prebuilt.d:
```diff
@@ -0,0 +1,313 @@
+}
+
+/// Expand the given string using BASH-like variable expansion.
+string expand(const(char)[] s)
```
Can we name this differently, s.t. it's not confused with the tuples expand?

marler8997 replied:
Renamed it to `shellExpand`.

marler8997 commented on test/run.d:
```diff
@@ -121,6 +121,7 @@ Options:
         auto taskPool = new TaskPool(jobs);
         scope(exit) taskPool.finish();
         ensureToolsExists(EnumMembers!TestTools);
+        buildDShellPrebuilt(env);
```
I could do that but the dshell_prebuilt object file is built by the `d_do_test` tool and therefore depends on it.  I initially tried having `run.d` build it rather than `d_do_test` but couldn't get it to work when I realized that having `d_do_test` build it would make more sense as it already has the right flags and environment because it's what actually builds the dshell tests.

The simplest way I can think to do that at the moment is add this to the end of ensureToolsExists

```
if (tool.name == "d_do_test")
{
    // now build dshell_prebuilt
    const testRunnerExe = resultsDir.buildPath(TestTools.testRunner.name).exeName;
    const command = [testRunnerExe, "-dshell_prebuilt"];
    spawnProcess(command, env, Config.none).wait;
}

It would also mean adding `env` as another argument to ensureToolsExist because it's required to know what flags to use when building the dshell_prebuilt object.

So long as you're ok with these things I'll change it.  Or if you have a better idea let me know.

marler8997 replied:
Nevermind, got `run.d` and `Makefile` to build it.  I think I just wasn't passing the whole environment.

marler8997 commented on test/tools/d_do_test.d:
```diff
@@ -576,8 +597,9 @@ int tryMain(string[] args)
         case "compilable":              testArgs.mode = TestMode.COMPILE;      break;
         case "fail_compilation":        testArgs.mode = TestMode.FAIL_COMPILE; break;
         case "runnable":                testArgs.mode = TestMode.RUN;          break;
+        case "dshell":                  testArgs.mode = TestMode.RUN;          break;
```
The dshell tests don't use or need any of the code below. Like the bash tests, we escape early into our own function because the logic to run them is completely different.  As you can see form the code, we do the same thing with the bash tests.

If we used a testMode instead of escaping early like this,, I'd have to disable alot of the code below behind something like `if (testMode != TestMode.dshell)` as it does things like tries to read test parameters from the input file.  And some test parameters aren't valid with dshell like DFLAGS, because if different flags are used to build the dshell test, then it wont' link to dshell_prebuilt anymore.  And there's no need to compile the dshell tests with different DFLAGS because, the point isn't to test the dshell tests, it's just to run them.

There would be some tests parameters that would make sense, like DISABLED_PLATFORMS, but that can already be done inside the test like the current bash tests do., i.e. `if (OS == "...")`.

marler8997 replied:
Oh wait, I think you just mean to add an enum value and assign it here....nevermind.  I've gone ahead and done that.

marler8997 commented on test/tools/d_do_test.d:
```diff
@@ -962,11 +970,145 @@ int runBashTest(string input_dir, string test_name)
+        exitCode = wait(runTestProc);
+    }
+
+    if (exitCode != 0)
```
Actually no, the exit code can be set inside the `if` block

marler8997 replied:
This logic is confusing.  I've put the code to print the test result into it's own function and now the 2 commands have their own `exitCode` variable.

marler8997 commented on test/tools/d_do_test.d:
```diff
@@ -962,11 +970,145 @@ int runBashTest(string input_dir, string test_name)
+}
+
+/// build the dshell prebuilt
+int buildDShellPrebuilt(const ref EnvData envData)
```
When I compiled `dshell_prebuilt` with `run.d` (which I started with) it doesn't link to the code compiled with `d_do_test`.  Since `d_do_test` is the one compiling the dshell tests, I realized it would be simpler to have `d_do_test` also compile `dshell_prebuilt` so we can ensure it compiles it in the same environment and with the same flags as the `dshell` tests themselves.

I could try again to get this to work with `run.d`, but I'll probably need to copy some code from `d_do_test`.  Also, the makefile will need to copy over whatever logic I put into `run.d`.  Right now they both just call `d_do_test -dshell_prebuilt` and all the logic is common and stays in `d_do_shell`.

marler8997 replied:
Same thing as the my previous comment.  This is no longer relevant as I got `run.d` and `Makefile` to build it directly now.

marler8997 commented on test/tools/dshell_prebuilt/dshell_prebuilt.d:
```diff
@@ -0,0 +1,313 @@
+void createFile(string filename, string contents)
+{
+    mkdirFor(filename);
+    auto file = File(filename, "w");
```
The problem is I don't remember the standard library as well as you :)

marler8997 commented:
>The idea was that you create a struct which has the internal variables, but instead of a > 100LOC function, there are a few smaller function which will drastically help a future reader to understand the code flow.

I believe doing that would make the code-flow harder to understand.  This may be a 100-line function but it has no loops and no midway jumps or returns.  Just a few if statements with a max depth of 2 and a few blocks to separate some variables into sub-scopes.  Having long functions is fine, it is more about what the function is doing I think.

By splitting the code into multiple functions now you have to start checking/verifying/understanding what the state in each method.  You can't assume that a function that should have run before ran unless you go through all the code and verify it is always called before you. It provides less guarantees as now you have to verify every place a function is called before you can know what state you can assume with the data it shares with the struct.

wilzbach commented on test/run.d:
```diff
@@ -147,26 +151,53 @@ void verifyCompilerExists(string[string] env)
-                sourceFile
-            ] ~ tool.extraArgs;
+            string[] command;
+            string[string]* commandEnv = null;
```
What's the motivation for making this a pointer to a pointer?
A D hashmap is already a pointer.

marler8997 replied:
Ah, for some reason I thought it had value semantics, but looks like I was wrong.

wilzbach commented on test/run.d:
```diff
@@ -147,26 +151,53 @@ void verifyCompilerExists(string[string] env)
 
             writefln("Executing: %-(%s %)", command);
-            spawnProcess(command).wait;
+            spawnProcess(command, commandEnv ? *commandEnv : null).wait;
```
```suggestion
            spawnProcess(command, commandEnv).wait;
```

wilzbach commented on test/run.d:
```diff
@@ -147,26 +151,53 @@ void verifyCompilerExists(string[string] env)
-        const sourceFile = toolsDir.buildPath(tool ~ ".d");
+        string targetBin;
+        string sourceFile;
+        if (tool.linksWithTests)
```
Maybe add a one-sentence here explaining sth. like this is preparing the shell tester D library  and thus needs the actual DMD compiler under test?

marler8997 replied:
I added a sentence where I add the compiler with `env["DMD"]`

Commit:
Start replacing bash tests with D

marler8997 commented:
@wilzbach suggested I try splitting up the `runDShellTest` function in `d_do_test.d`.  I did so but I think I still like the old way where it's just a function.  What do you guys think?

One Long Function:
https://github.com/marler8997/dmd/commit/b19751e851ef41ecef97ff23e1d5c6ab692f4266#diff-685230fdddcb87fcbb4a344b264a2fb6R994

Struct with Multiple SmallFunctions:
https://github.com/dlang/dmd/pull/9848/files#diff-685230fdddcb87fcbb4a344b264a2fb6R993

JinShil commented:
> What do you guys think?

JinShil commented:
> What do you guys think?

IMO, if a function does not have multiple callers, it should not be a function.  Rather, it should be surrounded in scope blocks `{ }` or separated by comment blocks with "weight".   D offers another option which is to use nested functions, which could be a nice compromise.  I'll take any of the presented options, though.

marler8997 commented:
> IMO, in general, if a function does not have multiple callers, it should not be a function. Rather, it should be surrounded in scope blocks { } or separated by comment blocks with "weight".

My preference as well.

jacob-carlborg commented:
> IMO, in general, if a function does not have multiple callers, it should not be a function. Rather, it should be surrounded in scope blocks { } or separated by comment blocks with "weight". 

That's how you end up with the massive functions that exists in DMD.

> D offers another option which is to use nested functions, which could be a nice compromise. I'll take any of the presented options, though.

That's a lot better. Wrapping the code in a function and giving it a name will automatically add some documentation.

marler8997 commented:
Here's the variant using nested functions: https://github.com/marler8997/dmd/commit/6889a74b63010d018d68d84b80e0562d2220f4a5#diff-685230fdddcb87fcbb4a344b264a2fb6R994

I like it better than the struct as it's closer to the one function design where there's less for the reader to keep track of.  But I still prefer just one function.  Adding these extra functions is strange to me since they're only called once, I'm expecting to find them being called somewhere else.  They do provide documentation, but comments and/or scopes can do the same thing without introducing the extra mental burden of having to track when these functions are called, so it confuses me when I see people introduce functions but only call them once because it makes me question why they would add the extra mental burden on the reader.

marler8997 commented:
Decided to go with the original one function solution.  The 3 who have weighed in seem to have different opinions with no one option winning out.

ibuclaw commented:
> @atilaneves another thing to consider is it looks like we've finished removing all the C++ code in DMD. Very similar to what this PR suggests except we're going from BASH to D. In fact, converting the backend to D has very similar pros and cons to converting the BASH scripts to D (except the performance reason).

The conversion from C++ to D had no refactoring of the original code during the process of translation.

If the translation from Shell to D does the same and makes sure there's a clear 1:1 mapping then it should be uncontroversial.

RazvanN7 commented:
What is the state of this? How should we move forward?

JinShil commented:
> What is the state of this?

In its current state, I don't think it can move forward as @atilaneves has firmly objected.

> How should we move forward?

IMO, the best way forward is to do each shell test one at time.   As @ibuclaw implied, to whatever extent possible, just do a direct translation from Bash to D.  If a few utility methods would make that easier and make the 1:1 mapping more apparent, so be it, but try to avoid creating too much supporting infrastructure.

wilzbach commented on test/tools/dshell_prebuilt/dshell_prebuilt.d:
```diff
@@ -0,0 +1,296 @@
+        return true;
+    if (c == '_')
+        return true;
+    return (c >= 'a' && c <= 'z');
```
Why not 

```d
import std.ascii : isAlphaNum;
return c.isAlphaNum || c == '_';
```

https://dlang.org/phobos/std_ascii.html#isAlphaNum

marler8997 replied:
Good suggestion. Fixed.

wilzbach reviewed (approved):
Minor remark, can be merged afterwards.

wilzbach commented:
> In its current state, I don't think it can move forward as @atilaneves has firmly objected.

I disagree, this is fine. We need to move forward here and @atilaneves objected with:

> I'm willing to do the work of getting the existing bash tests working on Windows.

which very likely won't materialize. This is a clear step forward and I strongly think we should take it.

atilaneves commented:
> which very likely won't materialize. This is a clear step forward and I strongly think we should take it.

It will. Unfortunately for me as soon as I tried I couldn't even get the master branch of dmd to compile, much less do anything else about it. And then I haven't had time since, but this clearly needs to happen.

atilaneves commented:
> > What is the state of this?
> 
> In its current state, I don't think it can move forward as @atilaneves has firmly objected.

I firmly objected to porting all the *existing* tests, not to this PR in particular. Writing new tests using this PR is probably the right thing to do.

JinShil commented:
> Unfortunately for me as soon as I tried I couldn't even get the master branch of dmd

If you were trying this on Windows, yes, it's a common problem and complaint.  Run `rdmd src/build.d` on Windows.  It will only work for DMD, though, as [we need to make a `build.d` conterpart for druntime and Phobos still](https://github.com/dlang/projects/issues/44).  Everything will just be smoother for everyone in the end if we use D for our infrastructure.

> Writing new tests using this PR is probably the right thing to do.

I'll take it.  As reviewers I hope we can begin expecting any new shell-like tests to be written in D.

12345swordy commented:
@thewilsonator 72 hours is up, is this ready to merge?

marler8997 commented:
Ready on my end
